Raven.Client.Lightweight/Connection/ReplicationInformer.cs

```diff
@@ -44,22 +44,29 @@ public ReplicationInformer(DocumentConvention conventions, HttpJsonRequestFactor
 
         public Task UpdateReplicationInformationIfNeededAsync(AsyncServerClient serverClient, bool force = false)
         {
-            return UpdateReplicationInformationIfNeededInternalAsync(serverClient.Url, () => 
-                AsyncHelpers.RunSync(() => serverClient.DirectGetReplicationDestinationsAsync(new OperationMetadata(serverClient.Url, serverClient.PrimaryCredentials, null), null)), force);
+            return UpdateReplicationInformationIfNeededInternalAsync(serverClient.Url,
+                () => serverClient.DirectGetReplicationDestinationsAsync(new OperationMetadata(serverClient.Url, serverClient.PrimaryCredentials, null), null), force);
         }
 
-        private Task UpdateReplicationInformationIfNeededInternalAsync(string url, Func<ReplicationDocumentWithClusterInformation> getReplicationDestinations, bool force)
+        private async Task UpdateReplicationInformationIfNeededInternalAsync(string url, Func<Task<ReplicationDocumentWithClusterInformation>> getReplicationDestinations, bool force)
         {
             if (force == false)
             {
                 if (Conventions.FailoverBehavior == FailoverBehavior.FailImmediately)
-                    return new CompletedTask();
+                    return;
 
                 if (lastReplicationUpdate.Add(Conventions.TimeToWaitBetweenReplicationTopologyUpdates) > SystemTime.UtcNow)
-                    return new CompletedTask();
+                    return;
             }
-            lock (replicationLock)
+
+            try
             {
+                // Do a quick synchronous check before we resort to async/await with the state-machine overhead.
+                if (!replicationLock.Wait(0))
```

Comment:
Previously this acquired a replication lock object while `RefreshReplicationInformationInternal` acquired `lock(this)` so I assumed it is OK to assume only one operation is ever allowed into this critical section

Reply:
Yes